src/OpenTelemetry.Exporter.Geneva/Internal/MsgPack/MsgPackTraceExporter.cs

```diff
@@ -320,71 +297,164 @@ internal static bool CacheIfPartOfHttpUrl(KeyValuePair<string, object?> entry, o
+        {
+            // if ResourceFieldNames is set, it overrides the existing prepopulated fields setting.
+            this.prepopulatedFields = [];
+        }
+
+        // this is guaranteed to not be null because it's set in the constructor
+        Guard.ThrowIfNull(this.prepopulatedFields);
+
+        foreach (var resourceAttribute in resourceAttributes)
         {
-            string key = entry.Key;
-            bool isDedicatedField = false;
+            var key = resourceAttribute.Key;
+            var value = resourceAttribute.Value;
 
-            if (entry.Value is string resourceValue)
+            var isWantedAttribute = false;
+            if (this.resourceFieldNames != null)
             {
-                switch (key)
+                // this might seem inefficient, but it's only run once and I don't expect there to be many resource attributes
+                foreach (var wantedAttribute in this.resourceFieldNames!)
                 {
-                    case "service.name":
-                        key = Schema.V40.PartA.Extensions.Cloud.Role;
-                        isDedicatedField = true;
-                        break;
-                    case "service.instanceId":
-                        key = Schema.V40.PartA.Extensions.Cloud.RoleInstance;
-                        isDedicatedField = true;
+                    if (wantedAttribute == key)
+                    {
+                        switch (value)
+                        {
+                            case bool:
+                            case byte:
+                            case sbyte:
+                            case short:
+                            case ushort:
+                            case int:
+                            case uint:
+                            case long:
+                            case ulong:
+                            case float:
+                            case double:
+                            case string:
+                                break;
+                            case null:
+                                // This should be impossible because Resource attributes cannot have null values.
+                                // But just in case, turn it into something serializable to avoid crashing.
+                                value = "<NULL>";
+                                break;
+                            default:
+                                // Try to construct a value that communicates that the type is not supported.
+                                try
+                                {
+                                    var stringValue = Convert.ToString(value, CultureInfo.InvariantCulture);
+                                    if (stringValue == null)
+                                    {
+                                        value = "<Unsupported type>";
+                                    }
+
+                                    value = $"<Unsupported type: {stringValue}>";
```

Comment:
The value here is always set using this line, so anything set beforehand will be overridden.

value = $"<Unsupported type: {stringValue}>";